velox/vector/tests/DecodedVectorTest.cpp

```diff
@@ -790,6 +802,224 @@ TEST_F(DecodedVectorTest, dictionaryWrapOnConstantVector) {
+    DecodedVector decodedVector(*dict);
+    EXPECT_FALSE(decodedVector.isConstantMapping());
+    EXPECT_TRUE(decodedVector.base() == intNullableVector.get());
+
+    // Now wrap intVector which does not have any nulls.
+    auto wrappedVector = decodedVector.wrap(intVector, *dict, vectorSize);
+    EXPECT_TRUE(isDictionary(wrappedVector->encoding()));
+    // Ensure all nulls are propagated correctly.
+    assertEqualVectors(dict, wrappedVector);
+  }
+
+  // Case 10: Dictionary2(Dictionary1(Flat))
+  // Dictionary1: adds nulls, Dictionary2: no nulls, Flat: has nulls
+  // verify that nulls from the base (innermost vector) are also propagated in
+  // the wrap.
+  {
+    // Flat vector identical intVector but has a null at index 1.
+    auto intNullableVector =
+        makeFlatVector<int32_t>(vectorSize, [](auto row) { return row; });
+    intNullableVector->setNull(1, true);
+    // Dict null at indices = 0, 2, 4
+    auto dict = BaseVector::wrapInDictionary(
+        nulls, indices, vectorSize, intNullableVector);
+    dict = BaseVector::wrapInDictionary(nullptr, indices, vectorSize, dict);
+    DecodedVector decodedVector(*dict);
+    EXPECT_FALSE(decodedVector.isConstantMapping());
+    EXPECT_TRUE(decodedVector.base() == intNullableVector.get());
+
+    // Now wrap intVector which does not have any nulls.
+    auto wrappedVector = decodedVector.wrap(intVector, *dict, vectorSize);
+    EXPECT_TRUE(isDictionary(wrappedVector->encoding()));
+    // Ensure all nulls are propagated correctly.
+    assertEqualVectors(dict, wrappedVector);
+  }
+
+  // Case 11: Dictionary(Flat)
+  // Dictionary: no nulls, Flat: has nulls
+  // Ideally the nulls from the base (innermost vector) should be propagated in
+  // the wrap. But a previous fix tried to skip adding nulls if only the base
+  // had nulls and the wrap did not have any nulls. This bug was introduced in
+  // #2678 as a solution to another bug caused due to how Cast Expr mis-uses the
+  // wrap functionality to implement its own peeling. Therefore, it should be
+  // reverted once #3553 is merged. Keeping this test case for now with a
+  // TODO to ensure it is updated with the right behavior once its fixed.
+  {
+    // Flat vector identical intVector but has a null at index 1.
+    auto intNullableVector =
+        makeFlatVector<int32_t>(vectorSize, [](auto row) { return row; });
+    intNullableVector->setNull(1, true);
+    auto dict = BaseVector::wrapInDictionary(
+        noNulls, indices, vectorSize, intNullableVector);
+    DecodedVector decodedVector(*dict);
+    EXPECT_FALSE(decodedVector.isConstantMapping());
+    EXPECT_TRUE(decodedVector.base() == intNullableVector.get());
+
+    // Now wrap intVector which does not have any nulls.
+    auto wrappedVector = decodedVector.wrap(intVector, *dict, vectorSize);
+    EXPECT_TRUE(isDictionary(wrappedVector->encoding()));
+    // TODO: Switch this to ensure all nulls from base are propagated correctly
```

Comment:
Please see description for Case 11 added after L994 above. This change was added to fix a bug in cast where decodedvector is used. Since we are not moving ahead with https://github.com/facebookincubator/velox/pull/3553 I will have to drill into the initial failure that resulted in #2678 and more sure it wont fail if i fix this here.